### PR TITLE
`includeHTML()` warns if complete document detected

### LINKED ISSUE #124: includeHTML() should at least warn against full HTML documents
I don't remember how many times I have explained to users why `includeHTML()` won't work if they include a full HTML document: https://github.com/rstudio/rmarkdown/issues/1514 (I have seen and explained it in private emails, Twitter, Stack Overflow, and Github, for use cases in R Markdown and Shiny)

I'd like `includeHTML()` to have an argument `force = FALSE`: when it detects `<html>` or `<body>` in the HTML document, it should signal an error. Of course, detection via regular expression is not robust, so in case of false alarms (which should be relatively rare), users can choose to use `force = TRUE`. It is extremely frustrating that `includeHTML()` is always silent, and the output HTML page looks weird or empty in the browser. This is very difficult for those who don't understand HTML to debug.

Given the enormous pain, I'll be happy to submit a PR :)

### PR DESCRIPTION
**Pull request by gadenbuie**
Fixes #124 by emitting a warning when `includeHTML()` is passed a path to a file containing a complete HTML document.

This PR originally contained an implementation of `includeHTMLDocument()`. I'll open a new issue to track progress on that front, but I didn't want to hold up this change on working out the details we'd need to address there.

## Example

`includeHTML()` now warns with a recommendation to use `includeHTMLDocument()`

``` r
# Create a temporary complete HTML file
tmp_html <- withr::local_tempfile(fileext = ".html")
save_html(p("Test"), tmp_html)

includeHTML(tmp_html) |> print(browse = FALSE)
#> Warning message:
#> `includeHTML()` was provided a `path` that appears to be a complete HTML document.
#> ✖ Path: /var/folders/yl/0xhq7c7n3z7_9hp1356kctbr0000gp/T//Rtmp9wJCqR/filee0569372d04.html
#> ℹ Use `tags$iframe()` to include an HTML document. You can either ensure `path`
#>   is accessible in your app or document (see e.g. `shiny::addResourcePath()`)
#>   and pass the relative path to the `src` argument. Or you can read the 
#>   contents of `path` and pass the contents to `srcdoc`. 
#> <!DOCTYPE html>
#> <html lang="en">
#> <head>
#> <meta charset="utf-8"/>
#> <style>body{background-color:white;}</style>
#> 
#> 
#> </head>
#> <body>
#> <p>Test</p>
#> </body>
#> </html>
```

### CODE REVIEW

File: R/tags.R [hunk 1]
```diff
@@ -1734,15 +1734,81 @@ knit_print.html_dependency <- knit_print.shiny.tag
...
+    frameborder = "0"
+  )
+
+  iframe_args <- utils::modifyList(iframe_args_default, dots_list(...))
+
+  if (inherits(path, "AsIs") || grepl("^http?s://", path)) {
+    iframe_args$src <- as.character(path)
+  } else {
+    lines <- readLines(path, warn=FALSE, encoding='UTF-8')
+    iframe_args$srcdoc <- paste8(lines, collapse='\n')
```
Comment (cpsievert) [inline R/tags.R:1806]: Let's make sure this works well with `save_html()` on a `{htmlwidgets}`

File: R/tags.R [hunk 2]
```diff
@@ -1734,15 +1734,81 @@ knit_print.html_dependency <- knit_print.shiny.tag
...
+
+  iframe_args_default <- list(
+    width = "100%",
+    height = "400px",
+    frameborder = "0"
+  )
+
+  iframe_args <- utils::modifyList(iframe_args_default, dots_list(...))
+
+  if (inherits(path, "AsIs") || grepl("^http?s://", path)) {
```
Comment (gadenbuie) [inline R/tags.R:1802]:
```suggestion
  if (inherits(path, "AsIs") || grepl("^https?://", path)) {
```

File: R/tags.R [hunk 3]
```diff
@@ -1734,15 +1734,81 @@ knit_print.html_dependency <- knit_print.shiny.tag
...
+    iframe_args$src <- as.character(path)
+  } else {
+    lines <- readLines(path, warn=FALSE, encoding='UTF-8')
+    iframe_args$srcdoc <- paste8(lines, collapse='\n')
+  }
+
+  return(tags$iframe(class = "html-fill-item", !!!iframe_args))
```
Comment (cpsievert) [inline R/tags.R:1809]:
```suggestion
  bindFillRole(tags$iframe(!!!iframe_args), fill = fill)
```

File: R/tags.R [hunk 4]
```diff
@@ -1734,15 +1734,81 @@ knit_print.html_dependency <- knit_print.shiny.tag
...
+#'   `<iframe>` element. When `path` is a local file, the contents of the file
+#'   will be embedded in the `<iframe>` using the `srcdoc` attribute, which
+#'   embeds the complete HTML document directly in the page where
+#'   `includeHTMLDocument()` was called. When `path` is detected to be a full
+#'   URL starting with `https://` or is wrapped in `I()`, the `src` attribute of
+#'   the `<iframe>` will be set to the URL.
+#'
+#' @rdname include
+#' @export
+includeHTMLDocument <- function(path, ...) {
```
Comment (cpsievert) [inline R/tags.R:1791]:
```suggestion
includeHTMLDocument <- function(path, ..., height = NULL, fill = is.null(height)) {
```

File: R/tags.R [hunk 5]
```diff
@@ -1734,15 +1734,81 @@ knit_print.html_dependency <- knit_print.shiny.tag
...
+    return(FALSE)
+  }
+
+  # There are more requirements for the HTML document to be technically complete
+  # and valid, but the above conditions are sufficient for us to warn that the
+  # document should not be treated as an HTML fragment.
+  TRUE
+}
+
+#' @note `includeHTMLDocument()` embeds the contents of the file in an
```
Comment (schloerke) [inline R/tags.R:1781]: What is the difference between `@note` and `@describeIn`? Could we use `@describeIn include DESC` here?

Comment (schloerke) [reply R/tags.R:1781]: I guess this would have to be applied to the whole file.

No need to change it, still curious about the difference though.

**Review by schloerke — APPROVED** [review]

**Comment by gadenbuie** [comment]
Thanks for the input, I'm going to move this to draft while I consider next steps.

**Comment by gadenbuie** [comment]
@cpsievert @schloerke I updated this PR to reduce the scope and address the immediate problem posed in #124. With this change, `includeHTML()` will issue a warning if it detects that the user is attempting to include a complete HTML document. 

I've opened a new issue #383 to track thoughts and progress toward a new `includeHTMLDocument()` function, which uses the implementation I've removed from this PR.

**Review by wch — APPROVED** [review]

**Comment by wch** [comment]
Looks good -- just needs a NEWS entry.

**Pull request merged**